tools/monorepo-utils/src/changelog/lib/__tests__/projects.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * External dependencies
+ */
+import path from 'path';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getAllProjectsPathsFromWorkspace,
+	getChangeloggerProjectPaths,
+	getTouchedChangeloggerProjectsPathsMappedToProjects,
+} from '../projects';
+
+const sampleWorkspaceYaml = `
+packages:
+    - 'folder-with-lots-of-projects/*'
+    - 'projects/cool-project'
+    - 'projects/very-cool-project'
+    - 'interesting-project'
+`;
+const tmpRepoPath = path.join( __dirname, 'test-repo' );
+
+describe( 'Changelog project functions', () => {
+	it( 'getAllProjectsPathsFromWorkspace should provide a list of all projects supplied by pnpm-workspace.yml', async () => {
+		const projects = await getAllProjectsPathsFromWorkspace(
+			tmpRepoPath,
+			sampleWorkspaceYaml
+		);
+		const expectedProjects = [
+			'folder-with-lots-of-projects/project-b',
+			'folder-with-lots-of-projects/project-a',
+			'projects/cool-project',
+			'projects/very-cool-project',
+			'interesting-project',
+		];
+
+		expectedProjects.forEach( ( expectedProject ) => {
+			expect( projects ).toContain( expectedProject );
+		} );
+
+		expect( projects ).toHaveLength( expectedProjects.length );
+	} );
+
+	it( 'getChangeloggerProjectPaths should provide a list of all projects that use Jetpack changelogger', async () => {
+		const projects = await getAllProjectsPathsFromWorkspace(
+			tmpRepoPath,
+			sampleWorkspaceYaml
+		);
+		const changeloggerProjects = await getChangeloggerProjectPaths(
+			tmpRepoPath,
+			projects
+		);
+
+		const expectedChangeLoggerProjects = [
+			'folder-with-lots-of-projects/project-b',
+			'folder-with-lots-of-projects/project-a',
+			'projects/very-cool-project',
+		];
+
+		expectedChangeLoggerProjects.forEach(
+			( expectedChangeLoggerProject ) => {
+				expect( projects ).toContain( expectedChangeLoggerProject );
+			}
+		);
+
+		expect( changeloggerProjects ).toHaveLength(
+			expectedChangeLoggerProjects.length
+		);
+	} );
+
+	it( 'getTouchedChangeloggerProjectsPathsMappedToProjects should combine touched and changelogger projects and return a list that is a subset of both', async () => {
+		const touchedFiles = [
+			'folder-with-lots-of-projects/project-b/src/index.js',
+			'projects/very-cool-project/src/index.js',
+		];
+		const changeLoggerProjects = [
+			'folder-with-lots-of-projects/project-b',
+			'folder-with-lots-of-projects/project-a',
+			'projects/very-cool-project',
+		];
+		const intersectedProjects =
+			getTouchedChangeloggerProjectsPathsMappedToProjects(
+				touchedFiles,
+				changeLoggerProjects
+			);
+
+		expect( intersectedProjects ).toHaveLength( 2 );
+		expect( intersectedProjects ).toContain(
+			'folder-with-lots-of-projects/project-b'
+		);
+		expect( intersectedProjects ).toContain( 'projects/very-cool-project' );
+	} );
+
+	it( 'getTouchedChangeloggerProjectsPathsMappedToProjects should map plugins and js packages to the correct name', async () => {
```

Comment:
In my personal nitpick mode this doesn't check whether paths are mapped to the correct project, as we're checking across four different projects here. Could be nice to add checks that ensure a direct mapping -- e.g., with files only touching a single project. (Say, `touchedWooCommerceFiles` touch only the `plugins/woocommerce` project, and `touchedBetaTesterFiles` only touch the `plugins/beta-tester` project.)